ui/js/components/auth.js

```diff
@@ -0,0 +1,165 @@
+                  data-toggle="dropdown" data-target="#"
+                  class="btn btn-view-nav btn-right-navbar nav-persona-btn">
+            <div class="nav-user-icon">
+              <span class="fa fa-user pull-left"></span>
+            </div>
+            <span class="fa fa-angle-down lightgray"></span>
+          </button>
+          <ul class="dropdown-menu" role="menu" aria-labelledby="logoutLabel">
+              <li>
+                <a ng-click="$ctrl.logout()">Logout</a>
+              </li>
+          </ul>
+        </span>
+
+        <a class="btn btn-view-nav btn-right-navbar nav-login-btn"
+           ng-if="!$ctrl.user.loggedin"
+           ng-click="$ctrl.login()">Login/Register</a>
+    `,
+    bindings: {
+        // calls to the HTML which sets the user value in the $rootScope.
+        onUserChange: "&"
+    },
+    controller: ['$location', '$window', 'localStorageService',
+        'ThUserModel', '$http', 'thUrl', '$timeout',
+        function ($location, $window, localStorageService,
+                  ThUserModel, $http, thUrl, $timeout) {
+            var ctrl = this;
+            ctrl.user = {};
+            // "clears out" the user when it is detected to be logged out.
+            var loggedOutUser = {is_staff: false, username: "", email: "", loggedin: false};
+
+            /**
+             * Using a window listener here because I wasn't getting reliable
+             * results from the events from angular-local-storage.
+             */
+            $window.addEventListener("storage", function(e) {
+                if (e.key === "treeherder.user") {
+                    $timeout(function() {
+                        var newUser = JSON.parse(e.newValue);
+                        if (newUser && newUser.email) {
+                            // User was saved to local storage. Use it.
+                            ctrl.setLoggedIn(newUser);
+                        } else {
+                            // This is a tiny hack.  :)
+                            // Right after login, sometimes we get another
+                            // event come in here with a null value for
+                            // ``newValue``.  This happens once every 10 or so
+                            // logins, in my experience.  But localStorage
+                            // is still set to the logged in user.  So we just
+                            // double-check the localStorage value and possibly
+                            // skip the logout.
+                            var storedUser = localStorageService.get("user");
+                            if (!storedUser || !storedUser.loggedin) {
+                                ctrl.setLoggedOut();
+                            }
+                        }
+                    }, 0);
+                }
+            });
+
+            // Ask the back-end if a user is logged in on page load
```

Comment:
This is great since it means if the Django session has expired since the last time the page was loaded (the session defaults to 2 weeks iirc), then we don't blindly use the user value from localstorage when they actually need to sign in again.

I was thinking it would probably be worth filing a new bug after this one, which moved this block to a separate function that could also be called by `ThModelErrors`, such that if we ever get a 401/403 from the API, we can double check whether they're still signed in, and update the state if they're not. (Currently I think they have to refresh the page manually).